### PR TITLE
asciidoctor: add rouge to dependencies

### PR DESCRIPTION
**Pull request by svalaskevicius**
###### Motivation for this change

Add syntax highlighting support via rouge.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/typesetting/asciidoctor/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Rebuild gemset.nix and Gemfile.lock with:
```
Comment (danbst) [inline pkgs/tools/typesetting/asciidoctor/README.md]: offtopic: to be pedantic, you should use .adoc here =)

**Comment by danbst** [comment]
I like that README is added (I've had this chicken-and-egg problem with `gemset.nix`). However I think it is enough to do:

```
$ nix-shell -p bundix --run 'bundix'
```
to rebuild `gemset.nix`.

Also, why should we add `rouge`?  There already is 'coderay', 'pygments.rb'. Which benefits has `rouge` compared to those?

**Comment by svalaskevicius** [comment]
I've tried pygments a while back with asciidoctor and it just didn't work.. Found this to be an easy workaround for what I needed instead of debugging pygments (or maybe failure to use it?) 

On a more global scale, adding it here costs little and gives users more choice what to use

**Comment by svalaskevicius** [comment]
Regarding bundix, I've added nix-shell to be able to regenerate it without having it installed.. As I stumbled on it a few times - otherwise it fails on missing dependencies

**Comment by danbst** [comment]
> As I stumbled on it a few times - otherwise it fails on missing dependencies.

I've tried `rm gemset.nix && nix-shell --pure -p bundix --run 'bundix'` and it was enough to regenerate `gemset.nix`. The `rebuild-gemset.nix` is redundant here, just mention the command I've proposed in README.adoc.

> On a more global scale, adding it here costs little and gives users more choice what to use

Ok. I don't know Ruby packaging details, I'm curious how it behaves when I use `asciidoctor` and `rouge` in local Gemfile/gemset.nix. I _hope_ there will be no conflicts.

File: pkgs/tools/typesetting/asciidoctor/rebuild-gemset.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{ pkgs ? import <nixpkgs> {} }:
+  pkgs.mkShell {
+# TODO: how to reuse the list from default.nix?
+    buildInputs = with pkgs; [
```
Comment (worldofpeace) [inline pkgs/tools/typesetting/asciidoctor/rebuild-gemset.nix]: I'd think with #62971 this would be uneeded

**Comment by svalaskevicius** [comment]
Thanks, I'll try reproducing the dependencies error when regenerating. Will report back here when done. Will update readme too.

**Comment by danbst** [comment]
@svalaskevicius have you reproduced the issue?

**Comment by svalaskevicius** [comment]
Sorry didn't have a chance yet. Will try to do it this evening :)

**Comment by svalaskevicius** [comment]
ok, reproduced, (@danbst) 

your command works ok only if Gemfile.lock is present, however, it doesn't add new libs to it, only converts the lock file to gemset.nix.

to add rouge to the lock file, one apparently needs to run `bundix -m`, which runs ruby things.. that fail to install mathematical because there are no binary dependencies installed:
```
<compiling things and dep errors>
An error occurred while installing mathematical (1.6.12), and Bundler cannot continue.
Make sure that `gem install mathematical -v '1.6.12' --source 'https://rubygems.org/'` succeeds before bundling.                                                             
```

Thus, I still don't see any other alternative rather than in this PR.

I'll update the readme to adoc though :)

File: pkgs/tools/typesetting/asciidoctor/rebuild-gemset.nix (hunk 2 above)
Comment (svalaskevicius) [reply pkgs/tools/typesetting/asciidoctor/rebuild-gemset.nix]: sorry, not quite sure what you mean :) do you mean this file, or this list here?

any chance you could update the PR to the preferred approach? (please also see my other comment on the PR where I failed to generate the gemset and Gemfile.lock in any other approach I've tried... )

**Comment by svalaskevicius** [comment]
done. also when testing now I noticed that rouge and some other dep has upped their version

**Comment by danbst** [comment]
@svalaskevicius thanks! Makes sense. I found a way to remove duplication. Here's diff:
```diff
diff --git a/pkgs/tools/typesetting/asciidoctor/default.nix b/pkgs/tools/typesetting/asciidoctor/default.nix
index 9508f3ff023..e46ebcf18ae 100644
--- a/pkgs/tools/typesetting/asciidoctor/default.nix
+++ b/pkgs/tools/typesetting/asciidoctor/default.nix
@@ -1,10 +1,10 @@
-{ stdenv, lib, bundlerApp, ruby
+{ stdenv, lib, bundlerApp, ruby, bundix, mkShell
   # Dependencies of the 'mathematical' package
 , cmake, bison, flex, glib, pkgconfig, cairo
 , pango, gdk_pixbuf, libxml2, python3, patchelf
 }:
 
-bundlerApp {
+bundlerApp rec {
   inherit ruby;
   pname = "asciidoctor";
   gemdir = ./.;
@@ -43,6 +43,12 @@ bundlerApp {
     };
   };
 
+  passthru.updateShell = mkShell {
+    buildInputs = (gemConfig.mathematical {}).buildInputs ++ [
+      bundix
+    ];
+  };
+
   meta = with lib; {
     description = "A faster Asciidoc processor written in Ruby";
     homepage = http://asciidoctor.org/;
diff --git a/pkgs/tools/typesetting/asciidoctor/update.sh b/pkgs/tools/typesetting/asciidoctor/update.sh
new file mode 100755
index 00000000000..10a053a847b
--- /dev/null
+++ b/pkgs/tools/typesetting/asciidoctor/update.sh
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+rm gemset.nix Gemfile.lock
+nix-shell ../../../.. -A asciidoctor.updateShell --run '
+  bundix -m --bundle-pack-path $TMPDIR/asciidoctor-ruby-bundle
+'
+rm -r .bundle
\ No newline at end of file
```
That can also remove the need for README, as `update.sh` is kinda standard way for package update scripts.

If you like it, you may embed it into this PR. Otherwise I can push it separately.

**Comment by svalaskevicius** [comment]
Thanks! I'll test and add it a bit later :)

**Comment by svalaskevicius** [comment]
works well! I've set you as the author in the commit @danbst if it's ok with you :)

**Comment by danbst** [comment]
@svalaskevicius thanks for doublechecking!

**Pull request merged**